<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.6: http://docutils.sourceforge.net/" />
<title>LAST: Genome-Scale Sequence Comparison</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7056 2011-06-17 10:50:48Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math {
  margin-left: 2em ;
  margin-right: 2em }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }
/* Style sheet for LAST HTML documents */
h1 { color: navy }
h2 { color: teal }
div.document { margin-left: auto; margin-right: auto; max-width: 40em }
li { margin-bottom: 1ex }

</style>
</head>
<body>
<div class="document" id="last-genome-scale-sequence-comparison">
<h1 class="title">LAST: Genome-Scale Sequence Comparison</h1>

<p>LAST finds similar regions between sequences, and aligns them.  It is
designed for comparing large datasets to each other (e.g. vertebrate
genomes and/or large numbers of DNA reads).  It can:</p>
<ul class="simple">
<li>Indicate the (un)ambiguity of each column in an alignment.</li>
<li>Use sequence quality data in a rigorous fashion.</li>
<li>Align DNA to proteins with frameshifts.</li>
<li>Compare PSSMs to sequences.</li>
<li>Calculate the likelihood of chance similarities between random
sequences.</li>
</ul>
<div class="section" id="requirements">
<h2>Requirements</h2>
<p>To handle mammalian genomes, it's best if you have at least 10-20
gigabytes of real memory, but you can get by with 2 gigabytes.</p>
<p>To install the software, you need a C++ compiler.  On Linux, you might
need to install a package called &quot;g++&quot;.  On Mac, you might need to
install command-line developer tools.  On Windows, you might need to
install Cygwin.</p>
</div>
<div class="section" id="setup">
<h2>Setup</h2>
<p>Using the command line, go into the top-level LAST directory.  To
compile the programs, type:</p>
<pre class="literal-block">
make
</pre>
<p>Optionally, copy the programs and scripts to a standard &quot;bin&quot;
directory (using &quot;sudo&quot; to request administrator permissions):</p>
<pre class="literal-block">
sudo make install
</pre>
<p>Or copy them to your personal bin directory:</p>
<pre class="literal-block">
make install prefix=~
</pre>
<p>You might have to log out and back in before your computer recognizes
the new programs.</p>
</div>
<div class="section" id="usage">
<h2>Usage</h2>
<p>Please see the other files in the doc directory, especially
<a class="reference external" href="last-tutorial.html">last-tutorial.html</a>.</p>
</div>
<div class="section" id="references">
<h2>References</h2>
<p>If you would like to understand LAST in detail, please look at these
articles.  If you use LAST in your research, please cite one (or more)
of them.</p>
<p>The main algorithms used by LAST:</p>
<blockquote>
Adaptive seeds tame genomic sequence comparison.
SM Kielbasa, R Wan, K Sato, P Horton, MC Frith: Genome Research 2011 21:487.</blockquote>
<p>How LAST uses sequence quality data:</p>
<blockquote>
Incorporating sequence quality data into alignment improves DNA read mapping.
MC Frith, R Wan, P Horton: Nucleic Acids Research 2010 38(7):e100.</blockquote>
<p>Choice of score parameters, ambiguity of alignment columns, and
gamma-centroid alignment:</p>
<blockquote>
Parameters for Accurate Genome Alignment.
MC Frith, M Hamada, P Horton: BMC Bioinformatics 2010 11:80.</blockquote>
</div>
<div class="section" id="other-credits">
<h2>Other credits</h2>
<ul>
<li><p class="first">LAST includes public domain code from Yi-Kuo Yu and Stephen Altschul
(NCBI), to infer probabilities from a score matrix.  This is used
for incorporating sequence quality data, and estimating column
ambiguity.  Reference:</p>
<blockquote>
<p>Yu Y-K, Altschul SF: The construction of amino acid substitution
matrices for the comparison of proteins with non-standard
compositions. Bioinformatics 21(2005), 902-911.</p>
</blockquote>
</li>
<li><p class="first">LAST includes public domain code from the Spouge group (NCBI), which
forms the basis of lastex.  Reference:</p>
<blockquote>
<p>S Sheetlin, Y Park, JL Spouge &quot;The Gumbel pre-factor k for gapped
local alignment can be estimated from simulations of global
alignment&quot; (2005) Nucl Acids Res 33: 4987-4994.</p>
</blockquote>
</li>
</ul>
</div>
<div class="section" id="license">
<h2>License</h2>
<p>LAST (including the scripts) is distributed under the GNU General
Public License, either version 3 of the License, or (at your option)
any later version.</p>
</div>
<div class="section" id="website">
<h2>Website</h2>
<p>LAST's website is: <a class="reference external" href="http://last.cbrc.jp/">http://last.cbrc.jp/</a></p>
</div>
<div class="section" id="contact">
<h2>Contact</h2>
<p>Questions and feedback are very welcome.  Please use the public
mailing list:</p>
<blockquote>
last-align (ATmark) googlegroups (dot) com</blockquote>
<p>Please let us know about any problems, rather than giving up in
disgust.  Feedback is essential for scientific software like this: we
cannot solve problems that we are unaware of, and we cannot make it
useful and convenient without learning how various people actually
(try to) use it.  It is also valuable to hear success stories, so we
know what we are doing right.  Unlike some other software projects, we
will never send rude or mocking replies.</p>
<p>Having said that, we cannot promise to always help.  If you find an
error, please tell us the exact error message, the exact commands that
trigger it, and any other potentially-relevant information (e.g. your
operating system).</p>
<p>If you do a benchmarking test of LAST, we recommend you contact us to
check you are using it in a suitable way.  The history of
bioinformatics benchmarks shows it is all too easy to get this wrong.</p>
</div>
</div>
</body>
</html>
