<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.6: http://docutils.sourceforge.net/" />
<title>maf-convert</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7056 2011-06-17 10:50:48Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math {
  margin-left: 2em ;
  margin-right: 2em }

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }
/* Style sheet for LAST HTML documents */
h1 { color: navy }
h2 { color: teal }
div.document { margin-left: auto; margin-right: auto; max-width: 40em }
li { margin-bottom: 1ex }

</style>
</head>
<body>
<div class="document" id="maf-convert">
<h1 class="title">maf-convert</h1>

<p>This script reads alignments in maf format, and writes them in another
format.  It can write them in these formats: axt, blast, html, psl,
sam, tab.  You can use it like this:</p>
<pre class="literal-block">
maf-convert psl my-alignments.maf &gt; my-alignments.psl
</pre>
<p>It's often convenient to pipe in the input, like this:</p>
<pre class="literal-block">
... | maf-convert psl &gt; my-alignments.psl
</pre>
<p>The input should be &quot;multiple alignment format&quot; as described in the
UCSC Genome FAQ (not &quot;MIRA assembly format&quot; or any other maf).</p>
<p>This script takes the first (topmost) MAF sequence as the &quot;reference&quot;
/ &quot;subject&quot; / &quot;target&quot;, and the second sequence as the &quot;query&quot;.</p>
<p>For html: if the input includes probability lines starting with 'p',
then the output will be coloured by column probability.  (To get lines
starting with 'p', run lastal with option -j set to 4 or higher.)</p>
<div class="section" id="options">
<h2>Options</h2>
<blockquote>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></td>
<td>Print a help message and exit.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-p</span>, <span class="option">--protein</span></kbd></td>
<td>Specify that the alignments are of proteins, rather than
nucleotides.  This affects psl format only (the first 4
columns).</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-n</span>, <span class="option">--noheader</span></kbd></td>
<td>Omit any header lines from the output.  This may be useful if
you concatenate outputs, e.g. from parallel jobs.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-d</span>, <span class="option">--dictionary</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Include a dictionary of sequence lengths in the sam header
section (lines starting with &#64;SQ).  This requires reading the
input twice, so it must be a real file (not a pipe).  This
affects sam format only.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-f <var>DICTFILE</var></span>, <span class="option">--dictfile=<var>DICTFILE</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Get a sequence dictionary from DICTFILE.  This affects sam
format only.  You can create a dict file using
CreateSequenceDictionary (<a class="reference external" href="http://picard.sourceforge.net/">http://picard.sourceforge.net/</a>).</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-r <var>READGROUP</var></span>, <span class="option">--readgroup=<var>READGROUP</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Specify read group information.  This affects sam format
only.  Example: -r 'ID:1 PL:ILLUMINA SM:mysample'</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-l <var>CHARS</var></span>, <span class="option">--linesize=<var>CHARS</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Write CHARS characters per line.  This affects blast and html
formats only.</td></tr>
</tbody>
</table>
</blockquote>
</div>
<div class="section" id="hints-for-sam-bam">
<h2>Hints for sam/bam</h2>
<ul>
<li><p class="first">To run fast on multiple CPUs, and get a correct header at the top,
this may be the least-awkward way.  First, make a header (perhaps by
using CreateSequenceDictionary).  Then, concatenate the output of a
command like this:</p>
<pre class="literal-block">
parallel-fastq &quot;... | maf-convert -n sam&quot; &lt; q.fastq
</pre>
</li>
<li><p class="first">Here is yet another way to get a sequence dictionary, using samtools
(<a class="reference external" href="http://samtools.sourceforge.net/">http://samtools.sourceforge.net/</a>).  Assume the reference sequences
are in ref.fa.  These commands convert x.sam to y.bam while adding a
sequence dictionary:</p>
<pre class="literal-block">
samtools faidx ref.fa
samtools view -bt ref.fa.fai x.sam &gt; y.bam
</pre>
</li>
<li><p class="first">If a query name ends in &quot;/1&quot; or &quot;/2&quot;, maf-convert interprets it as a
paired sequence.  (This affects sam format only.)  However, it does
not calculate all of the sam pairing information (because it's hard
and better done by specialized sam manipulators).</p>
<p>Fix the pair information in y.sam, putting the output in z.bam.
Using picard:</p>
<pre class="literal-block">
java -jar FixMateInformation.jar I=y.sam O=z.bam VALIDATION_STRINGENCY=SILENT
</pre>
<p>Alternatively, using samtools:</p>
<pre class="literal-block">
samtools sort -n y.bam ysorted
samtools fixmate ysorted.bam z.bam
</pre>
</li>
</ul>
</div>
<div class="section" id="bugs">
<h2>&quot;Bugs&quot;</h2>
<ul class="simple">
<li>For sam: the QUAL field (column 11) is simply copied from the maf q
line.  The QUAL is supposed to be encoded as ASCII(phred+33),
whereas maf q lines are encoded differently according to the UCSC
Genome FAQ.  However, if you run lastal with option -Q1, the maf q
lines will in fact be ASCII(phred+33).</li>
<li>The blast format is merely blast-like: it is not identical to NCBI
BLAST.</li>
</ul>
</div>
</div>
</body>
</html>
